classlib/src/main/java/org/teavm/classlib/java/lang/TString.java

```diff
@@ -418,20 +423,20 @@ public int lastIndexOf(TString str) {
     }
 
     public TString substring(int beginIndex, int endIndex) {
-        if (beginIndex > endIndex) {
-            throw new TIndexOutOfBoundsException();
+        int length = charactersLength();
+        if (beginIndex < 0 || beginIndex > endIndex || endIndex > length) {
```

Comment:
I think this check is only necessary for JS. In other BEs TString constructor will perform the same check